File: cplusplus/546.cpp
#include <iostream>
#include <cstdlib>
#include <stack>
#include <cstdio>
#include <vector>
#include <string>
#include <cstring>
#include <algorithm>
#include <map>
#include <climits>
#include "./utils.h"


class Solution {
public:
    int dp[100][100][100];

    int removeBoxes(vector<int>& boxes) {
        memset(dp, 0, sizeof dp);
        return calculatePoints(boxes, 0, boxes.size() - 1, 0);
    }

    int calculatePoints(vector<int>& boxes, int l, int r, int k) {
        printf("%d %d %d\n", l, r, k);
        if (l > r) return 0;
        if (dp[l][r][k] != 0) return dp[l][r][k];
        while (r > l && boxes[r] == boxes[r - 1]) {
            r--;
            k++;
        }
        dp[l][r][k] = calculatePoints(boxes, l, r - 1, 0) + (k + 1) * (k + 1);
        for (int i = l; i < r; i++) {
            if (boxes[i] == boxes[r]) {
                dp[l][r][k] = max(dp[l][r][k], calculatePoints(boxes, l, i, k + 1) + calculatePoints(boxes, i + 1, r - 1, 0));
            }
        }
        // printf("%d %d %d %d\n", l, r, k, dp[l][r][k]);
        printf("   %d\n", dp[l][r][k]);
        return dp[l][r][k];
    }
};


int main(int argc, const char** argv) {
    // vector<int> list = {1,3,2,2,2,3,4,3,1};
    vector<int> list = {1,3,2,2,3};
    // printVector(list);
    printf("%d\n", Solution().removeBoxes(list));
    return 0;
}

File: python/546.py
class Solution:
    def removeBoxes(self, boxes):
        res = 0
        size = len(boxes)
        if size == 0:
            return res
        dp = [[[0]*100]*100]*100
        def solve(l, r, k):
            print(l,r,k)
            nonlocal dp
            if l > r:
                return 0
            if dp[l][r][k] != 0:
                return dp[l][r][k]
            
            while r>l and boxes[r]==boxes[r-1]:
                r -= 1
                k += 1

            ans = solve(l, r-1, 0) + (k+1)*(k+1)
            for i in range(l, r):
                if boxes[i] == boxes[r]:
                    ans = max(ans, solve(l, i, k+1) + solve(i+1, r-1, 0))
            dp[l][r][k] = ans
            print('  ', dp[l][r][k])
            return dp[l][r][k]
        return solve(0, size-1, 0)

# boxes = [1,3,2,2,2,3,4,3,1]
boxes = [1,3,2,2,3]
print(Solution().removeBoxes(boxes))